### PR TITLE
Profile dynamo guards

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #96199
* __->__ #96119

Adds a profiler start and end callback to dynamo's C eval_frame impl, which can be used to profile a region providing a name for visualization.  Currently only hooks up one usage to profile cache lookup (primarily covering guards and linear search through  linked list).
 
Example profile taken from toy model:
`python benchmarks/dynamo/distributed.py --toy_model --profile --dynamo aot_eager`
image

Planning to measure overhead in CI, and probably can't afford to check this in enabled by default.  Will have to evaluate UX options such as `config.profile_dynamo_cache = True` or some other way.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire @ezyang @mlazos

### CODE REVIEW

File: torch/csrc/dynamo/eval_frame.c [hunk 1]
```diff
@@ -476,6 +479,26 @@ static PyObject* call_guard_fail_hook(
...
+  return result;
+}
+
+static void call_profiler_end_hook(PyObject* record) {
+  // 'record' obj is the return value of calling _start_hook()
+  if (profiler_end_hook == NULL) return;
+  if (record == NULL) return;
+  PyObject* args = PyTuple_Pack(1, record);
+  if (args == NULL) return;
+  PyObject_CallObject(profiler_end_hook, args);
```
Comment (jansel) [inline torch/csrc/dynamo/eval_frame.c:498]: Memory leak?  Return value is owned, but you don't decrement refcount.

File: torch/csrc/dynamo/eval_frame.c [hunk 2]
```diff
@@ -640,7 +663,10 @@ static PyObject* _custom_eval_frame(
   // we never compile.
   if (callback == Py_False) {
     DEBUG_TRACE("In run only mode %s", name(frame));
+    PyObject* hook_record = call_profiler_start_hook(guard_profiler_name_str);
```
Comment (jansel) [inline torch/csrc/dynamo/eval_frame.c:666]: Memory leak?  Return hook_record is owned, but you don't decrement refcount.

File: torch/csrc/dynamo/eval_frame.c [hunk 3]
```diff
@@ -840,13 +868,47 @@ static PyObject* set_guard_error_hook(PyObject* dummy, PyObject* args) {
...
+  if (start == Py_None || end == Py_None) {
+    profiler_start_hook = NULL;
+    profiler_end_hook = NULL;
+  } else {
+    profiler_start_hook = start;
+    profiler_end_hook = end;
+    Py_INCREF(profiler_start_hook);
+    Py_INCREF(profiler_end_hook);
+  }
+  guard_profiler_name_str = Py_BuildValue("s", "TorchDynamo Cache Lookup");
```
Comment (jansel) [inline torch/csrc/dynamo/eval_frame.c:888]: Memory leak if there is an existing value?

File: torch/csrc/dynamo/eval_frame.c [hunk 4]
```diff
@@ -840,13 +868,47 @@ static PyObject* set_guard_error_hook(PyObject* dummy, PyObject* args) {
...
+  PyObject* start = NULL;
+  PyObject* end = NULL;
+  if (!PyArg_ParseTuple(args, "OO", &start, &end)) {
+    return NULL;
+  }
+  Py_XDECREF(profiler_start_hook);
+  Py_XDECREF(profiler_end_hook);
+  if (start == Py_None || end == Py_None) {
+    profiler_start_hook = NULL;
+    profiler_end_hook = NULL;
```
Comment (jansel) [inline torch/csrc/dynamo/eval_frame.c:881]: Call clear_profiler_hooks to reuse code?

**Review by jansel — CHANGES_REQUESTED** [review]

**Comment by weiwangmeta** [comment]
FYI, https://github.com/pytorch/pytorch/pull/95685 was merged therefore the ciflow/inductor-perf-test-nightly label will now trigger a 10hour+ job (instead of ~1hour) on this PR, similar to this one https://github.com/pytorch/pytorch/actions/runs/4337030784

**Comment by wconstab** [comment]
> FYI, https://github.com/pytorch/pytorch/pull/95685 was merged therefore the ciflow/inductor-perf-test-nightly label will now trigger a 10hour+ job (instead of ~1hour) on this PR, similar to this one https://github.com/pytorch/pytorch/actions/runs/4337030784

@weiwangmeta thanks for the warning but i'm not sure what you meant.
 - is there a new label for a faster job now?
 - or is this a temporary thing for some one-off effort to collect data
 - or perf CI is just changed now?

**Comment by weiwangmeta** [comment]
> > FYI, #95685 was merged therefore the ciflow/inductor-perf-test-nightly label will now trigger a 10hour+ job (instead of ~1hour) on this PR, similar to this one https://github.com/pytorch/pytorch/actions/runs/4337030784
> 
> @weiwangmeta thanks for the warning but i'm not sure what you meant.
> 
> * is there a new label for a faster job now?
> * or is this a temporary thing for some one-off effort to collect data
> * or perf CI is just changed now?

cc @desertfire for help answering some of the questions. My understanding is this perf CI job is changed now to be large/slow job, running once everyday for 10+ hours. It is not a temporary thing. 
We talked about adding a faster job too.

**Comment by weiwangmeta** [comment]
ha, looks like you can avoid the 10hours issue by not rebasing :)

File: torch/csrc/dynamo/eval_frame.c (hunk 1 above)
Comment (wconstab) [reply torch/csrc/dynamo/eval_frame.c:498]: good catch, thanks!

**Review by jansel — APPROVED** [review]

**Comment by wconstab** [comment]
pasting perf analysis- saw minor degradation as expected across hf/timm (~0.5-1% geomean), with a notable outlier of  DeBerta models regressing a lot more.  

Suspect many graph breaks in those models led to increased impact of small profiler overhead 
`python benchmarks/dynamo/huggingface.py --only DebertaForMaskedLM --accuracy --inductor --training --explain`
-->  "Dynamo produced 53 graphs covering 712 ops with 51 graph breaks (11 unique)"
